src/main/java/pl/edu/agh/firecell/engine/algorithm/FirePropagation.java

```diff
@@ -0,0 +1,163 @@
+                        MAX_BURNING_TIME != 0 &&
+                        oldCell.flammable()) {
+                    newBurningTime ++;
+                }
+                if (oldCell.burningTime() > 0 && oldCell.burningTime() <= MAX_BURNING_TIME) {
+                    newBurningTime++;
+                }
+                if (oldCell.burningTime() > MAX_BURNING_TIME) {
+                    newBurningTime = 0;
+                }
+            }
+            case AIR -> {
+                newBurningTime = 0;
+
+                if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                        oldState.getCell(NeighbourUtils.down(cellIndex)).burningTime() > 0 &&
+                        oldState.getCell(NeighbourUtils.down(cellIndex)).flammable() &&
+                        oldState.getCell(NeighbourUtils.down(cellIndex)).remainingFirePillar() - 1 > 0) {
+                    newBurningTime++;
+                }
+
+                if (!oldState.hasCell(NeighbourUtils.down(cellIndex)) ||
+                        oldState.getCell(NeighbourUtils.down(cellIndex)).burningTime() == 0 ||
+                        !oldState.getCell(NeighbourUtils.down(cellIndex)).flammable()) {
+
+                    Optional<Cell> neighbourOnFire = doesNeighbourOnFire(oldState, cellIndex);
+                    if (neighbourOnFire.isPresent() &&
+                            !neighbourOnFire.get().possibleToGoUp() &&
+                            neighbourOnFire.get().remainingFirePillar() - 1 > 0)
+                    {
+                        newBurningTime++;
+                    }
+                }
+            }
+        }
+        return newBurningTime;
+    }
+
+    private static Optional<Cell> doesNeighbourOnFire(State oldState, Vector3i cellIndex){
+        try {
+            Cell northCell = oldState.getCell(NeighbourUtils.north(cellIndex));
+            Cell southCell = oldState.getCell(NeighbourUtils.south(cellIndex));
+            Cell westCell = oldState.getCell(NeighbourUtils.west(cellIndex));
+            Cell eastCell = oldState.getCell(NeighbourUtils.east(cellIndex));
+            if (doesHorizontalFire(northCell))
+                return Optional.of(northCell);
+            if (doesHorizontalFire(southCell))
+                return Optional.of(southCell);
+            if (doesHorizontalFire(westCell))
+                return Optional.of(westCell);
+            if (doesHorizontalFire(eastCell))
+                return Optional.of(eastCell);
+        } catch (IndexOutOfBoundsException ignored) {}
+        return Optional.empty();
+    }
+
+    private static boolean doesHorizontalFire(Cell cell){
+        return cell.burningTime() > 0 && cell.remainingFirePillar() > 0 && !cell.possibleToGoUp();
+    }
+
+    public int computeFirePillar(State oldState, Cell oldCell, Vector3i cellIndex, int currentFirePillar){
+        int newRemainingHeightOfFirePillar = currentFirePillar;
+        if(oldCell.material().equals(AIR)){
```

Comment:
Don't create if { if { if { tower. Just check if material is not air and return.